File: 297.serialize-and-deserialize-binary-tree.cpp
/*
 * @lc app=leetcode id=297 lang=cpp
 *
 * [297] Serialize and Deserialize Binary Tree
 */
#include <string>
#include <queue>

using std::stoi;
using std::string;
using std::to_string;
using std::queue;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    // Encodes a tree to a single string.
    // T: O(n)
    // S: O(n) recursion stack worst case
    string serialize(TreeNode* root) {
        // Pre-order traversal
        string res;
        build_str(root, res);
        return res;
    }

    // Decodes your encoded data to tree.
    // T: O(n)
    // S: O(n)
    TreeNode* deserialize(string data) {
        // extract nodes (incl. null)
        queue<string> q;
        string tmp;
        for (const auto& c : data) {
            if (c == ',') {
                // A new node or subtree
                q.push(std::move(tmp));
                // tmp.clear();
                continue;
            }
            tmp.push_back(c);
        }
        q.push(std::move(tmp));  // don't forget the last one
        
        return build_tree(q);
    }

private:
    void build_str(TreeNode* node, string& res) {
        if (node == nullptr) {
            res.push_back('X');
            return;
        }

        // Pre-order traversal
        res.append(to_string(node->val)).push_back(',');
        build_str(node->left, res);
        res.push_back(',');  // IMPORTANT separate subtrees
        build_str(node->right, res);
    }

    TreeNode* build_tree(queue<string>& q) {
        // Pre-order traversal
        const string val = q.front(); q.pop();
        if (val == "X") {
            return nullptr;
        }
        TreeNode* node = new TreeNode(stoi(val));
        node->left = build_tree(q);
        node->right = build_tree(q);
        return node;
    }

};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));
// @lc code=end

File: 297.serialize-and-deserialize-binary-tree.java
/*
 * @lc app=leetcode id=297 lang=java
 *
 * [297] Serialize and Deserialize Binary Tree
 */
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {
    private static final String NULL_NODE_STR = "X", SEPARATOR = ",";

    // Encodes a tree to a single string.
    // T: O(n)
    // S: O(n) recursion stack worst case
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        buildString(root, sb);
        return sb.toString();
    }

    // Pre-order traversal
    // T: O(n)
    // S: O(n) recursion stack worst case
    private void buildString(TreeNode node, StringBuilder sb) {
        if (node == null) {
            sb.append(NULL_NODE_STR);
        } else {
            sb.append(node.val).append(SEPARATOR);
            buildString(node.left, sb);
            sb.append(SEPARATOR);  // IMPORTANT
            buildString(node.right, sb);
        }
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        Queue<String> q = new ArrayDeque<>(Arrays.asList(data.split(SEPARATOR)));
        return buildTree(q);
    }

    private TreeNode buildTree(Queue<String> q) {
        String val = q.poll();
        if (val.equals(NULL_NODE_STR)) return null;
        TreeNode root = new TreeNode(Integer.valueOf(val));
        root.left = buildTree(q);
        root.right = buildTree(q);
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));
// @lc code=end

